app/coffee/Features/Collaborators/CollaboratorsInviteHandler.coffee

```diff
@@ -0,0 +1,118 @@
+				callback(null, invite)
+
+	revokeInvite: (projectId, inviteId, callback=(err)->) ->
+		logger.log {projectId, inviteId}, "removing invite"
+		ProjectInvite.remove {projectId: projectId, _id: inviteId}, (err) ->
+			if err?
+				logger.err {err, projectId, inviteId}, "error removing invite"
+				return callback(err)
+			callback(null)
+
+	getInviteByToken: (projectId, tokenString, callback=(err,invite)->) ->
+		logger.log {projectId, tokenString}, "fetching invite by token"
+		ProjectInvite.findOne {projectId: projectId, token: tokenString}, (err, invite) ->
+			if err?
+				logger.err {err, projectId}, "error fetching invite"
+				return callback(err)
+			if !invite
+				logger.err {err, projectId, token: tokenString}, "no invite found"
+				return callback(null, null)
+			callback(null, invite)
+
+	acceptInvite: (projectId, inviteId, tokenString, user, callback=(err)->) ->
+		# fetch the target project
+		Project.findOne {_id: projectId}, (err, project) ->
+			if err?
+				logger.err {err, projectId}, "error finding project"
+				return callback(err)
+			if !project
+				err = new Errors.NotFoundError("no project found for invite")
+				logger.log {err, projectId, inviteId}, "no project found"
+				return callback(err)
+			# fetch the invite
+			ProjectInvite.findOne {_id: inviteId, projectId: projectId, token: tokenString}, (err, invite) ->
+				if err?
+					logger.err {err, projectId, inviteId}, "error finding invite"
+					return callback(err)
+				if !invite
+					err = new Errors.NotFoundError("no matching invite found")
+					logger.log {err, projectId, inviteId, tokenString}, "no matching invite found"
+					return callback(err)
+
+				# check if user is already a member of this project,
+				# return early if so
+				existing_users = (project.collaberator_refs or [])
+				existing_users = existing_users.concat(project.readOnly_refs or [])
+				existing_users = existing_users.map (u) -> u.toString()
+				if existing_users.indexOf(user._id.toString()) > -1
+					logger.log {projectId, userId: user._id}, "user already member of project, returning"
+					return callback null
+
+				# build an update to be applied with $addToSet, user is added to either
+				# `collaberator_refs` or `readOnly_refs`
+				privilegeLevel = invite.privileges
+				if privilegeLevel == PrivilegeLevels.READ_AND_WRITE
+					level = {"collaberator_refs": user._id}
+					logger.log {privileges: privilegeLevel, user_id: user._id, projectId}, "adding user with read-write access"
+				else if privilegeLevel == PrivilegeLevels.READ_ONLY
+					level = {"readOnly_refs": user._id}
+					logger.log {privileges: privilegeLevel, user_id: user._id, projectId}, "adding user with read-only access"
+				else
+					return callback(new Error("unknown privilegeLevel: #{privilegeLevel}"))
+
+				ContactManager.addContact invite.sendingUserId, user._id
+
+				# Update the project, adding the new member.
+				Project.update { _id: project._id }, { $addToSet: level }, (error) ->
```

Comment:
All updates to the project collaborators should go through CollaboratorHandler. C.f. https://github.com/sharelatex/web-sharelatex/blob/master/app/coffee/Features/Collaborators/CollaboratorsHandler.coffee#L113